File: path_generator/src/path_generator.cpp
#include <ros/ros.h>
#include <nav_msgs/Path.h>
#include <geometry_msgs/PoseStamped.h>
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <cmath>

int main(int argc, char** argv)
{
  ros::init(argc, argv, "path_generator");
  
  ros::NodeHandle n;
  ros::Publisher path_pub = n.advertise<nav_msgs::Path>("path_pub", 1000);
  ros::Publisher pos_pub = n.advertise<geometry_msgs::PoseStamped>("pos_pub", 1000);

  int freq = 100;
  float round_min = 0.25;
  ros::Rate loop_rate(freq);
  
  nav_msgs::Path path_msg;
  path_msg.poses = {};
  
  float radius = 1.0;

  int index = 0;
  int count = 0;
  while (ros::ok())
  {
    count += 1;
    geometry_msgs::PoseStamped pos_msg;
    
    pos_msg.pose.position.x = radius * cos((M_PI * 2) / freq * round_min * count);
    pos_msg.pose.position.y = radius * sin((M_PI * 2) / freq * round_min * count);
    pos_msg.pose.position.z = 1.0;
    
    if (count <= (freq / round_min)) 
    {
      index = count;
      geometry_msgs::PoseStamped temp [index];
      for (int i = 0; i < index - 1; i++)
	temp[i] = path_msg.poses[i];
      temp[index - 1] = pos_msg;
      path_msg.poses = temp;
    }
    else
    {
      geometry_msgs::PoseStamped temp [index];
      for (int i = 0; i < index - 1; i++)
	temp[i] = path_msg.poses[i + 1];
      temp[index - 1] = pos_msg;
      path_msg.poses = temp;
    }

    // ROS_INFO("TEST");

    path_pub.publish(path_msg);
    pos_pub.publish(pos_msg);

    ros::spinOnce();

    loop_rate.sleep();
  }

  return 0;
}

File: path_generator/src/path_generator.py
#!/usr/bin/env python
from __future__ import print_function

import roslib
roslib.load_manifest('path_generator')
import rospy
import numpy as np
from nav_msgs.msg import Path
from geometry_msgs.msg import PoseStamped

def talker():
    path_pub = rospy.Publisher('/ardrone/path_pub', Path, queue_size=10)
    pos_pub = rospy.Publisher('/ardrone/pos_pub', PoseStamped, queue_size=10)
    rospy.init_node('path_generator', anonymous=True)
    
    freq = 100;
    round_min = 0.1;
    
    path_msg = Path();
    path_msg.header.stamp = rospy.get_rostime();
    path_msg.poses = [];
    path_msg.header.frame_id = "nav";
    
    radius = 2.0;    
    rate = rospy.Rate(freq) # 100hz

    count = 0;
    while not rospy.is_shutdown():
        count += 1;
        pos_msg = PoseStamped();
        pos_msg.header.stamp = rospy.get_rostime();
        pos_msg.header.frame_id = "nav";
        
        pos_msg.pose.position.x = radius * np.cos((np.pi * 2) / freq * round_min * count);
        pos_msg.pose.position.y = radius * np.sin((np.pi * 2) / freq * round_min * count);
        pos_msg.pose.position.z = 1.5;       
        
        if count > (freq / round_min):
            path_msg.poses.pop(0)
        path_msg.poses.append(pos_msg);    
        
        path_pub.publish(path_msg)
        pos_pub.publish(pos_msg)
        rate.sleep()

if __name__ == '__main__':
    try:
        talker()
    except rospy.ROSInterruptException:
        pass
